ui/src/app/components/popups.tsx

```diff
@@ -0,0 +1,285 @@
+  const getPosition = s => {
+    const left = flow(
+      clamp(0, viewport.width - element.width),
+      clamp(target.left - element.width + 16, target.right - 16)
+    )(((target.left + target.right) / 2) - (element.width / 2));
+    const top = flow(
+      clamp(0, viewport.height - element.height),
+      clamp(target.top - element.height + 16, target.bottom - 16)
+    )(((target.top + target.bottom) / 2) - (element.height / 2));
+    return switchCase(s,
+      ['top', () => ({ top: target.top - element.height - gap, left })],
+      ['bottom', () => ({ top: target.bottom + gap, left })],
+      ['left', () => ({ left: target.left - element.width - gap, top })],
+      ['right', () => ({ left: target.right + gap, top })]
+    );
+  };
+  const position = getPosition(side);
+  const maybeFlip = d => {
+    return switchCase(d,
+      ['top', () => position.top < 0 ? 'bottom' : 'top'],
+      ['bottom', () => position.top + element.height >= viewport.height ? 'top' : 'bottom'],
+      ['left', () => position.left < 0 ? 'right' : 'left'],
+      ['right', () => position.left + element.width >= viewport.width ? 'left' : 'right']
+    );
+  };
+  const finalSide = maybeFlip(side);
+  const finalPosition = getPosition(finalSide);
+  return { side: finalSide, position: finalPosition };
+};
+
+export const PopupPortal = ({ children }) => {
+  return createPortal(React.Children.only(children), document.getElementById('popup-root'));
+};
+
+export const Tooltip = withDynamicPosition()(class TooltipComponent extends React.Component {
+  static defaultProps = {
+    side: 'bottom'
+  };
+
+  props: any;
+
+  render() {
+    const {children, side, elementRef, dimensions: {target, element, viewport}} = this.props;
+    const {side: finalSide, position} =
+      computePopupPosition({ side, target, element, viewport, gap: 10 });
+    const getNotchPosition = () => {
+      const left = clamp(12, element.width - 12,
+        (target.left + target.right) / 2 - position.left
+      );
+      const top = clamp(12, element.height - 12,
+        (target.top + target.bottom) / 2 - position.top
+      );
+      return switchCase(finalSide,
+        ['top', () => ({ bottom: 0, left, transform: 'rotate(180deg)' })],
+        ['bottom', () => ({ top: 0, left })],
+        ['left', () => ({ right: 0, top, transform: 'rotate(90deg)' })],
+        ['right', () => ({ left: 0, top, transform: 'rotate(270deg)' })]
+      );
+    };
+    return <PopupPortal>
+      <div
+        ref={elementRef}
+        style={{
+          transform: `translate(${position.left}px, ${position.top}px)`,
+          ...styles.tooltip
+        } as React.CSSProperties}
+      >
+        {children}
+        <svg viewBox='0 0 2 1' style={{...getNotchPosition(), ...styles.notch}}>
+          <path d='M0,1l1,-1l1,1Z' />
+        </svg>
+      </div>
+    </PopupPortal>;
+  }
+});
+
+export class TooltipTrigger extends React.Component {
+  props: any;
```

Comment:
We should consider whether we want typedefs for props/state. If we decide to continue with Typescript, we should likely be typing these as they are pretty fundamental to coding a react component.

Reply:
Correct me if I am wrong @panentheos, but it looks like (from the defaultProps) thing, that these are pretty flexible/mutable? So it might be hard to pin down a typedef

Reply:
React itself doesn't care what you put in props, but any well-designed component should be expecting a particular structure, so you should be able to specify types if you want to. Same goes for state.